graphqlapi/src/main/java/io/stargate/graphql/schema/NameConversions.java

```diff
@@ -0,0 +1,240 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.graphql.schema;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+
+public class NameConversions {
+
+  enum IdentifierType {
+    TABLE,
+    UDT,
+    COLUMN,
+  }
+
+  private NameConversions() {}
+
+  public static String toGraphql(String cqlName, IdentifierType type) {
+    if (cqlName == null || cqlName.isEmpty()) {
+      throw new IllegalArgumentException("CQL name must be non-null and not empty");
+    }
+    if (type == IdentifierType.TABLE) {
+      for (Map.Entry<String, String> entry : RESERVED_GRAPHQL_NAMES.entrySet()) {
+        if (cqlName.equals(entry.getKey())) {
+          return entry.getValue();
+        }
+      }
+    }
+    String result = hexEscape(cqlName);
+    if (type == IdentifierType.TABLE) {
+      result = escapeReservedPrefix(result);
+      result = escapeReservedSuffix(result);
+    } else if (type == IdentifierType.UDT) {
+      result = escapeReservedPrefix(result);
+      // No need to escape suffixes because we append our own:
+      result += "Udt";
+    }
+    return result;
+  }
+
+  public static String toCql(String graphqlName, IdentifierType type) {
+    if (graphqlName == null || graphqlName.isEmpty()) {
+      throw new IllegalArgumentException("GraphQL name must be non-null and not empty");
+    }
+    if (type == IdentifierType.UDT) {
+      if (!graphqlName.endsWith("Udt")) {
+        throw new IllegalArgumentException("GraphQL UDT name must end with 'Udt'");
+      }
+      graphqlName = graphqlName.substring(0, graphqlName.length() - 3);
+    }
+    return unHexEscape(graphqlName);
+  }
+
+  private static String hexEscape(String source) {
+    StringBuilder result = null;
+    int length = source.length();
+    int i = 0;
+    while (i < length) {
+      char c = source.charAt(i);
+      // Handle this case separately because it consumes two characters:
+      if (i < length - 1 && Character.isSurrogatePair(c, source.charAt(i + 1))) {
+        result = appendHex(result, source, i);
+        i += 2;
+      }
+      // Other cases where we escape: leading double underscores, leading digit, character not
+      // allowed in GraphQL, or 'x' followed by something that could be interpreted as a hex escape.
+      else if ((i == 0 && length > 1 && c == '_' && source.charAt(1) == '_')
+          || (i == 0 && isDigit(c))
+          || !isValidGraphql(c)
+          || (c == 'x' && lengthOfHexEscape(source, i) > 0)) {
+        result = appendHex(result, source, i);
```

Comment:
I'm wondering if it would be clearer/faster to do a pre-pass checking for invalid chars instead of interleaving the allocation logic.

```java
    boolean invalidPrefix = hasInvalidPrefix(source);

    if (invalidPrefix || requiresHexEscape(source)) {
      StringBuilder result = new StringBuilder();

      // ...

      return result.toString();
    } else {
      return source;
    }
```

Reply:
Yeah my approach is a bit convoluted, the only advantage is that it traverses the string only once, but that's probably not significant enough to justify the complexity.
I'll do it as you suggest, thanks.

Reply:
Addressed in latest commit, it looks much cleaner now 👍